### ISSUE TITLE
msGEOSError(): GEOS library error when using MULTILINESTRING

### ISSUE THREAD

**Issue by gertcuykens** [mapserver/mapserver#5529]
msGEOSError(): GEOS library error when using MULTILINESTRING in shapefiles

Try `shp2img -m test.map -o test.png -l test -all_debug 5 -s 500 500`
on any shapefile that has MULTILINESTRING in it

example shapefile feature and using mapserver 7.0.7
```
OGRFeature(test):178

  color (String) = #c6b952

  MULTILINESTRING ((1365999.913051389390603
6668002.688304415903986,1365929.670452698832378
6667981.874043367803097),(1366036.091885895933956
6668063.352393131703138,1365999.913051389390603
6668002.688304415903986),(1366004.811108984053135
6668127.219178147614002,1366036.091885895933956
6668063.352393131703138),(1365934.123232331126928
6668138.427049811929464,1366004.811108984053135
6668127.219178147614002),(1365862.878758222097531
6668117.612443232908845,1365934.123232331126928
6668138.427049811929464),(1365792.190881569404155
6668094.485164736397564,1365862.878758222097531
6668117.612443232908845),(1365384.093628322007135
6667942.914158684201539,1365306.949221200309694
6667931.172859975136817),(1365461.460674423491582
6667952.342789551243186,1365384.093628322007135
6667942.914158684201539),(1365538.939040013123304
6667961.771431289613247,1365461.460674423491582
6667952.342789551243186),(1365616.528725096955895
6667969.065293673425913,1365538.939040013123304
6667961.771431289613247),(1365773.266568133374676
6668049.120308448560536,1365760.910104657290503
6668084.344763254746795),(1365771.819414753001183
6668087.724895685911179,1365792.190881569404155
6668094.485164736397564),(1365760.910104657290503
6668084.344763254746795,1365771.819414753001183
6668087.724895685911179),(1365694.786327125271782
6667973.512773931957781,1365616.528725096955895
6667969.065293673425913),(1365773.043929153587669
6667977.426558562554419,1365694.786327125271782
6667973.512773931957781),(1365851.301531179109588
6667980.272948558442295,1365773.043929153587669
6667977.426558562554419),(1365929.670452698832378
6667981.874043367803097,1365851.301531179109588
6667980.272948558442295))
```

**Comment by sdlime** [comment]
Can you share your test files? --Steve

**Comment by gertcuykens** [comment]
https://drive.google.com/file/d/1DC2XXjwbeA2DBexJUOgo-0nO9ZopMwXw/view?usp=sharing

**Comment by gertcuykens** [comment]
Note that i tried wgs84 and mercator, make sure to remove the prj file in the example and check ogrinfo because i tried so many things to get rid of it :)

**Comment by gertcuykens** [comment]
https://drive.google.com/file/d/1EGh9jwfNhoEW2f50LYiNnKpGYbms7UrD/view?usp=sharing

**Comment by gertcuykens** [comment]
Let me know if you can reproduce thx :)

**Comment by jratike80** [comment]
I had a try with MapServer version 7.0.6 (MS4W 3.2.2). Command
`shp2img -m trips.map -o test.png -l trips -all_debug 5 -s 500 500`
is printing lots of messages "msGEOSError(): GEOS library error. ýçè"
and the result is funny (shp2img.png).
![wms](https://user-images.githubusercontent.com/1751612/33714705-8d838894-db58-11e7-96fd-784fccbb1242.png)
![shp2img](https://user-images.githubusercontent.com/1751612/33714706-8e00eaa0-db58-11e7-874c-fbf691b61230.png)


However, if I use the layer as a WMS layer with OpenJUMP the rendering looks about good (wms.png)

**Comment by gertcuykens** [comment]
Correct render looks good, but it's the error message that doesn't make sense and will get send to a stderr on a nginx mapserver which is really annoying if you want to have some basic monitoring

**Comment by jratike80** [comment]
This is probably related with #5510. There is also negative offset in your mapfile
```
    CLASS
      STYLE
        width 8
        color "#000000"
        offset 4 -99
```
If I edit the offset into 4 99 the GEOS errors disappear. Shp2img creates still an image with points, not with lines.

**Comment by gertcuykens** [comment]
Aaa thank you, that i would never found myself :D Ps but it's a bug right? You should be able to use -99  right? as stated in the documentation? If I am not mistaken when using LINESTRING for example with the same map file structure -99 is valid.

**Comment by jratike80** [comment]
Usage is documented in http://www.mapserver.org/mapfile/style.html 

> For lines, an OFFSET of y = -99 will produce a line geometry that is shifted x SIZEUNITS perpendicular to the original line geometry. A positive x shifts the line to the right when seen along the direction of the line. A negative x shifts the line to the left when seen along the direction of the line.

If it does not work of if it just writes loads of geos errors into the log I would say that it is a bug.

**Comment by gertcuykens** [comment]
Link should be availeble again

**Comment by jbo-ads** [comment]
@gertcuykens Thank you, I downloaded the test data.

**Comment by gertcuykens** [comment]
@jbo-ads Ok suggest you also delete your 2 msg like i am going to do with this one to keep the issue report short and clean :) thx

**Comment by jbo-ads** [comment]
I am digging up this issue because a user I know is facing a similar bug. Here is what I've found so far.

Offset lines are computed with GEOS method `BufferBuilder::bufferLineSingleSided()` which requires a *LineString*, not a *MultiLineString*, see [code](https://git.osgeo.org/gitea/geos/geos/src/tag/3.8.1/src/operation/buffer/BufferBuilder.cpp#L129-L139) (call sequence is `maputil.c:msOffsetPolyline()` -> `maputil.c:msOffsetCurve()` -> `mapgeos.c:msGEOSOffsetCurve()` -> `geos/capi/geos_ts_c.cpp:GEOSOffsetCurve_r()` -> `src/operation/buffer/BufferBuilder.cpp:BufferBuilder::bufferLineSingleSided()`).

I am working on a fix inside `mapgeos.c:msGEOSOffsetCurve()` which basically consists in splitting a *MultilineString* into several *LineStrings* and compute offset on each *LineString*. Pull request is on the way.

**Comment by jmckenna** [comment]
I have confirmed @jbo-ads's fix here on Windows, with master & GEOS 3.8.0   The map image is generated, without any "`GEOS library error`" messages.

Thanks @jbo-ads and @gertcuykens !!
